core/instantiate.ml

```diff
@@ -400,57 +412,28 @@ let replace_quantifiers t qs' =
           `ForAll (Types.box_quantifiers qs', apply_type t tyargs)
     | t -> t
 
+let recursive_application name qs tyargs body =
+  let tenv, renv, penv = populate_instantiation_maps name qs tyargs in
+  let (_, body) = typ (instantiate_datatype (tenv, renv, penv) body) in
+  body
+
 let alias name tyargs env =
   (* This is just type application.
 
      (\Lambda x1 ... xn . t) (t1 ... tn) ~> t[ti/xi]
   *)
   match (SEnv.find env name : Types.tycon_spec option) with
     | None ->
-        failwith (Printf.sprintf "Unrecognised type constructor: %s" name)
-    | Some (`Abstract _) ->
-        failwith (Printf.sprintf "The type constructor: %s is abstract, not an alias" name)
+        internal_error (Printf.sprintf "Unrecognised type constructor: %s" name)
+    | Some (`Abstract _)
+    | Some (`Mutual _) ->
+        internal_error (Printf.sprintf "The type constructor: %s is not an alias" name)
     | Some (`Alias (vars, _)) when List.length vars <> List.length tyargs ->
-        failwith (Printf.sprintf
-                    "Type alias %s applied with incorrect arity (%d instead of %d)"
-                    name (List.length tyargs) (List.length vars))
+        internal_error (Printf.sprintf
```

Comment:
This fixes #184 by the way.  Actually a lot of lingering error-message issues may be subissues of #525.

Reply:
Hehe, splendid! I love accidentally-fixed issues.